### PR TITLE
GH703: Add FileLogger switches to MSBuild Runner

### PR DESCRIPTION
**Pull request by bradtwurst**
Addresses issue #703

Provides the ability to add FileLogger parameters to the MsBuild command
line.  The various fileloggerparameters (flp) can also be specified.

The parameters that can be created are /fl, /fl[1 thru 9], /flp, and
/flp[1 thru 9]

### CODE REVIEW

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 1]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
                              "\"/Working/src/Solution.sln\"", result.Args);
             }
+
+            [Fact]
+            public void Should_Append_FileLogger_To_Process_Arguments()
+            {
+                // Given
+                var fixture = new MSBuildRunnerFixture(false);
+                fixture.Settings.AddFileLogger(new MSBuildFileLogger { AppendToLogFile = false, Encoding = "E", HideVerboseItemAndPropertyList = false, LogFile = "A", MSBuildFileLoggerOutput = MSBuildFileLoggerOutput.All, PerformanceSummaryEnabled = false, ShowCommandLine = false, ShowEventId = false, ShowTimestamp = false, SummaryDisabled = false, Verbosity = Verbosity.Diagnostic });
+
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Remove line break.

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 2]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
+                fixture.Settings.AddFileLogger(new MSBuildFileLogger { AppendToLogFile = true, HideVerboseItemAndPropertyList = true, MSBuildFileLoggerOutput = MSBuildFileLoggerOutput.ErrorsOnly, PerformanceSummaryEnabled = true, ShowCommandLine = true, ShowEventId = true, ShowTimestamp = true, SummaryDisabled = true, Verbosity = Verbosity.Minimal });
+
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Remove line break.

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 3]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
+                fixture.Settings.AddFileLogger(new MSBuildFileLogger { MSBuildFileLoggerOutput = MSBuildFileLoggerOutput.WarningsOnly, Verbosity = Verbosity.Normal });
+
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Remove line break.

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 4]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
+                fixture.Settings.AddFileLogger(new MSBuildFileLogger { Verbosity = Verbosity.Quiet });
+
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Remove line break.

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 5]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
+                fixture.Settings.AddFileLogger(new MSBuildFileLogger { Verbosity = Verbosity.Verbose });
+
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Remove line break.

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 6]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
+                fixture.Settings.AddFileLogger(new MSBuildFileLogger { });
+
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Remove line break.

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 7]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
+                fixture.Settings.AddFileLogger();
+
+                // When
+                var result = fixture.Run();
+                // Then
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Add line break.

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 8]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
+                Assert.Equal(@"/v:normal /target:Build /fl /flp:logfile=A;Encoding=E;Verbosity=Diagnostic /fl1 /flp1:Append;PerformanceSummary;NoSummary;ErrorsOnly;NoItemAndPropertyList;ShowCommandLine;ShowTimestamp;ShowEventId;Verbosity=Minimal /fl2 /flp2:WarningsOnly;Verbosity=Normal /fl3 /flp3:Verbosity=Quiet /fl4 /flp4:Verbosity=Verbose /fl5 /fl6 ""/Working/src/Solution.sln""", result.Args);
+            }
+
+            [Fact]
+            public void Should_Append_Default_FileLogger_To_Process_Arguments()
+            {
+                // Given
+                var fixture = new MSBuildRunnerFixture(false);
+
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Remove line break.

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 9]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
+                fixture.Settings.AddFileLogger();
+
+                // When
+                var result = fixture.Run();
+                // Then
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Add line break.

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 10]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
+                Assert.Equal(@"/v:normal /target:Build /fl ""/Working/src/Solution.sln""", result.Args);
+            }
+
+            [Fact]
+            public void Should_Throw_Exception_For_Too_Many_FileLoggers()
+            {
+                // Given
+                var fixture = new MSBuildRunnerFixture(false);
+
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Remove line break.

File: src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs [hunk 11]
```diff
@@ -802,6 +802,69 @@ public void Should_Append_Logger_To_Process_Arguments()
...
+                fixture.Settings.AddFileLogger();
+                fixture.Settings.AddFileLogger();
+                fixture.Settings.AddFileLogger();
+                fixture.Settings.AddFileLogger();
+                fixture.Settings.AddFileLogger();
+                fixture.Settings.AddFileLogger();
+
+                // When
+                var ex = Assert.Throws<System.InvalidOperationException>(() => fixture.Run());
+                // Then
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Tools/MSBuild/MSBuildRunnerTests.cs]: Add line break.

File: src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        /// </summary>
+        public bool PerformanceSummaryEnabled { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether Summary will Show the error and warning summary at the end.
+        /// </summary>
+        public bool SummaryDisabled { get; set; }
+
+        /// <summary>
+        /// Gets or sets show ErrorsOnly, WarningsOnly, or All
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs]: Missing period.

File: src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        public bool ShowTimestamp { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether ShowEventId. Show the event ID for each started event, finished event, and message.
+        /// </summary>
+        public bool ShowEventId { get; set; }
+
+        /// <summary>
+        /// Gets or sets Verbosity. Override the /verbosity setting for this logger.
+        ///
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs]: Remove line break.

File: src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        /// specify the following verbosity levels: q[uiet], m[inimal], n[ormal], v[erbose] (detailed), and diag[nostic].
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs]: `Specify` should begin with capital letter.

File: src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+        /// </summary>
+        public Verbosity? Verbosity { get; set; }
+
+        /// <summary>
+        /// Gets or sets LogFile. The path to the log file into which the build log is written.
+        ///
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs]: Remove line break.

File: src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        /// an empty string will use msbuild.log
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        /// </summary>
+        public bool AppendToLogFile { get; set; }
+
+        /// <summary>
+        /// Gets or sets Specifies the encoding for the file (for example, UTF-8, Unicode, or ASCII).
+        /// </summary>
+        public string Encoding { get; set; }
+
+        /// <summary>
+        /// process the file logger config and return parameters as a string
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildFileLogger.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildFileLoggerOutput.cs [hunk 18]
```diff
@@ -0,0 +1,23 @@
+﻿namespace Cake.Common.Tools.MSBuild
+{
+    /// <summary>
+    /// the type of file logger output to generate
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildFileLoggerOutput.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildFileLoggerOutput.cs [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+    /// </summary>
+    public enum MSBuildFileLoggerOutput
+    {
+        /// <summary>
+        /// show errors and warnings
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildFileLoggerOutput.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildFileLoggerOutput.cs [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+        /// </summary>
+        All = 0,
+
+        /// <summary>
+        /// show errors only
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildFileLoggerOutput.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildFileLoggerOutput.cs [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+        /// </summary>
+        ErrorsOnly = 1,
+
+        /// <summary>
+        /// show warnings only
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildFileLoggerOutput.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildRunner.cs [hunk 22]
```diff
@@ -111,12 +111,44 @@ private ProcessArgumentBuilder GetArguments(FilePath solution, MSBuildSettings s
...
+                }
+            }
+
             // Add the solution as the last parameter.
             builder.AppendQuoted(solution.MakeAbsolute(_environment).FullPath);
 
             return builder;
         }
 
+        private static string GetLoggerArgument(int indx, MSBuildFileLogger logger)
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildRunner.cs]: `indx` should be called `index`

File: src/Cake.Common/Tools/MSBuild/MSBuildRunner.cs [hunk 23]
```diff
@@ -111,12 +111,44 @@ private ProcessArgumentBuilder GetArguments(FilePath solution, MSBuildSettings s
...
+        {
+            if (indx >= 10)
+            {
+                throw new InvalidOperationException("Too Many FileLoggers");
+            }
+
+            var cntr = indx == 0 ? string.Empty : indx.ToString();
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildRunner.cs]: `cntr` should be called `counter`.

File: src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs [hunk 24]
```diff
@@ -200,5 +200,50 @@ public static MSBuildSettings WithLogger(this MSBuildSettings settings, string l
             });
             return settings;
         }
+
+        /// <summary>
+        /// Adds a file logger.
+        ///
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs]: Remove line break.

File: src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs [hunk 25]
```diff
@@ -200,5 +200,50 @@ public static MSBuildSettings WithLogger(this MSBuildSettings settings, string l
             });
             return settings;
         }
+
+        /// <summary>
+        /// Adds a file logger.
+        ///
+        /// each file logger will be declared in the order added
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs [hunk 26]
```diff
@@ -200,5 +200,50 @@ public static MSBuildSettings WithLogger(this MSBuildSettings settings, string l
             });
             return settings;
         }
+
+        /// <summary>
+        /// Adds a file logger.
+        ///
+        /// each file logger will be declared in the order added
+        /// the first file logger will match up to the /fl parameter
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs [hunk 27]
```diff
@@ -200,5 +200,50 @@ public static MSBuildSettings WithLogger(this MSBuildSettings settings, string l
             });
             return settings;
         }
+
+        /// <summary>
+        /// Adds a file logger.
+        ///
+        /// each file logger will be declared in the order added
+        /// the first file logger will match up to the /fl parameter
+        /// the next nine (max) file loggers will match up to the /fl1 through /fl9 respectively
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs [hunk 28]
```diff
@@ -200,5 +200,50 @@ public static MSBuildSettings WithLogger(this MSBuildSettings settings, string l
...
+        /// <param name="settings">The settings.</param>
+        /// <param name="fileLoggerParameters">Parameters to be passed to the logger.</param>
+        /// <returns>The same <see cref="MSBuildSettings"/> instance so that multiple calls can be chained.</returns>
+        public static MSBuildSettings AddFileLogger(this MSBuildSettings settings, MSBuildFileLogger fileLoggerParameters)
+        {
+            if (settings == null)
+            {
+                throw new ArgumentNullException(nameof(settings));
+            }
+
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs]: Remove line break.

File: src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs [hunk 29]
```diff
@@ -200,5 +200,50 @@ public static MSBuildSettings WithLogger(this MSBuildSettings settings, string l
...
+            if (fileLoggerParameters == null)
+            {
+                throw new ArgumentNullException(nameof(fileLoggerParameters));
+            }
+            settings.FileLoggers.Add(fileLoggerParameters);
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs]: Add line break.

File: src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs [hunk 30]
```diff
@@ -200,5 +200,50 @@ public static MSBuildSettings WithLogger(this MSBuildSettings settings, string l
...
+            return settings;
+        }
+
+        /// <summary>
+        /// Adds a file logger with all the default settings.
+        ///
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs]: Remove line break.

File: src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs [hunk 31]
```diff
@@ -200,5 +200,50 @@ public static MSBuildSettings WithLogger(this MSBuildSettings settings, string l
...
+        /// each file logger will be declared in the order added
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs [hunk 32]
```diff
@@ -200,5 +200,50 @@ public static MSBuildSettings WithLogger(this MSBuildSettings settings, string l
...
+        /// the first file logger will match up to the /fl parameter
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs]: Sentence should begin with a capital letter and end with a period.

File: src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs [hunk 33]
```diff
@@ -200,5 +200,50 @@ public static MSBuildSettings WithLogger(this MSBuildSettings settings, string l
...
+        /// the next nine (max) file loggers will match up to the /fl1 through /fl9 respectively
```
Comment (patriksvensson) [inline src/Cake.Common/Tools/MSBuild/MSBuildSettingsExtensions.cs]: Sentence should begin with a capital letter and end with a period.

**Comment by gep13** [comment]
@bradtwurst I have rebased and merged this PR, and also added the PR Review comments from @patriksvensson. You can see this merged in here: https://github.com/cake-build/cake/commit/3dbf5101ae6ffc77e9d5f458970e366094c27af2

As a result, I am going to manually close this PR, but your contribution has been pulled in, and correctly attributed. Thank you for taking the time to contribute to cake! 👍

**Pull request closed without merging**